Address ipcache startup perfomance regression

bimmlerd opened the pull request:
The commit messages should be descriptive, here's the summary line for each.

1. ipcache: add benchmark for Upsert (`regression.txt` below)
2. ipcache: make NamedPortMultiMap an interface (`interface.txt` below)
3. ipcache: switch named ports to reference counting (`fix.txt` below)
4. ipcache: simplify the named ports update code (`cleanup.txt` below)

The following is a comparison of running the benchmark added in commit 1 over the course of the four commits.

```
goos: linux
goarch: arm64
pkg: github.com/cilium/cilium/pkg/ipcache
                      │ regression.txt │             interface.txt             │               fix.txt               │             cleanup.txt             │
                      │     sec/op     │     sec/op      vs base               │   sec/op     vs base                │   sec/op     vs base                │
IPCacheUpsert10-10        97.10µ ±  3%     94.58µ ±  3%       ~ (p=0.052 n=10)   48.84µ ± 5%  -49.70% (p=0.000 n=10)   47.59µ ± 5%  -50.98% (p=0.000 n=10)
IPCacheUpsert100-10      2672.3µ ±  5%    2718.4µ ±  8%       ~ (p=0.481 n=10)   508.3µ ± 5%  -80.98% (p=0.000 n=10)   485.0µ ± 2%  -81.85% (p=0.000 n=10)
IPCacheUpsert1000-10     69.543m ± 13%    66.515m ± 17%       ~ (p=0.481 n=10)   5.282m ± 5%  -92.40% (p=0.000 n=10)   5.056m ± 2%  -92.73% (p=0.000 n=10)
IPCacheUpsert10000-10   4166.59m ±  1%   4299.95m ±  2%  +3.20% (p=0.000 n=10)   48.63m ± 5%  -98.83% (p=0.000 n=10)   44.87m ± 5%  -98.92% (p=0.000 n=10)
geomean                   16.56m           16.47m        -0.55%                  1.589m       -90.40%                  1.513m       -90.86%
```

The last commit, `cleanup.txt` removes some now redundant work, and hence gets a mild speedup. If preferred, it can be squashed into the third commit, but like this the algorithmic changes are easier to understand.

```
goos: linux
goarch: arm64
pkg: github.com/cilium/cilium/pkg/ipcache
                      │   fix.txt   │            cleanup.txt             │
                      │   sec/op    │   sec/op     vs base               │
IPCacheUpsert10-10      48.84µ ± 5%   47.59µ ± 5%       ~ (p=0.165 n=10)
IPCacheUpsert100-10     508.3µ ± 5%   485.0µ ± 2%  -4.58% (p=0.005 n=10)
IPCacheUpsert1000-10    5.282m ± 5%   5.056m ± 2%  -4.28% (p=0.011 n=10)
IPCacheUpsert10000-10   48.63m ± 5%   44.87m ± 5%  -7.74% (p=0.009 n=10)
geomean                 1.589m        1.513m       -4.81%
```

Note that simply reverting 93cd67f2994a618865fbcdb0294bcd9b4d44036a is not actually a fix, as the quadratic behaviour still lurks. This can be observed by forcefully setting `ipcache.needNamedPorts = true` when benchmarking:

```
                      │ main-head.txt  │    revert-with-neednamedports.txt     │ 
                      │     sec/op     │     sec/op      vs base               │  
IPCacheUpsert10-10        95.00µ ±  5%     92.12µ ±  3%  -3.04% (p=0.000 n=10)   
IPCacheUpsert100-10      2729.7µ ±  4%    2617.5µ ± 12%       ~ (p=0.165 n=10)   
IPCacheUpsert1000-10     70.192m ± 11%    68.974m ± 10%       ~ (p=0.579 n=10)   
IPCacheUpsert10000-10   4247.52m ±  3%   4233.62m ±  1%       ~ (p=0.853 n=10)   
geomean                   16.68m           16.29m        -2.31%                 
```

Fixes: #24987 

```release-note
Address cilium-agent startup performance regression.
```

bimmlerd committed:
ipcache: add benchmark for Upsert

This patch adds a benchmark for ipcache.Upsert, with the intention of
exercising the logic for named ports. This exposes a performance
regression when upserting many pods, as is the case on agent startup.

The startup regression was due to losing an optimization in 93cd67f299,
namely that the named ports bookkeeping is done only once named ports
are present in policies. Without the lazy bookkeeping, we have
accidentally quadratic behavior, as we look at every pod whenever a
new pod is added.

Follow up commits will address the issue, this is merely a convenient
reproduction of the regression.

Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

christarazi commented on pkg/ipcache/ipcache_test.go:
```diff
@@ -553,6 +556,83 @@ func (s *IPCacheTestSuite) TestIPCacheNamedPorts(c *C) {
+	benchmarkIPCacheUpsert(10000, b)
+}
+
+func benchmarkIPCacheUpsert(num int, b *testing.B) {
```
Nit: Maybe `b` should be the first argument in case we need to adjust this in the future, it makes more sense to have `b` be first and the rest of the params follow

christarazi commented on pkg/ipcache/ipcache_test.go:
```diff
@@ -553,6 +556,83 @@ func (s *IPCacheTestSuite) TestIPCacheNamedPorts(c *C) {
+			DatapathHandler:   &mockTriggerer{},
+			NodeHandler:       &mockNodeHandler{},
+		})
+		b.StartTimer()
```
Nit: Why not put the `StartTimer` at the start of the for loop iteration and keep the `StopTimer` where it is? Seems a bit strange to put the start at the bottom of the loop iteration, even though the iteration will start back up again when the loop reaches the end.

christarazi commented on pkg/types/portmap.go:
```diff
@@ -31,7 +33,7 @@ type PortProto struct {
 
 // PortProtoSet is a set of unique PortProto values.
-type PortProtoSet map[PortProto]struct{}
+type PortProtoSet counter.Counter[PortProto]
```
What's the plan for the backport as this uses generics?

bimmlerd replied:
I was thinking we can use something akin to https://github.com/cilium/cilium/blob/v1.12/pkg/counter/string.go. It'll be yet another custom-ish backport, but I think it's worth having the nicer version on main.

christarazi replied:
I've marked the PR with `backport/author` for you to take care of the backport due to the above.

christarazi reviewed (commented):
Thanks for the fix, nice work! The fix looks good to me, I have a few minor questions below. 

Just a nit on the commit msg: technically the namedport map is not the only lock being held while it's locked; the ipcache lock is always held before the namedport map lock is taken. Actually, it seems like we should "promote" this relationship to say that the ipcache lock must be taken before the namedport map.

gandro commented on pkg/ipcache/ipcache.go:
```diff
@@ -244,44 +244,6 @@ func (ipc *IPCache) getK8sMetadata(ip string) *K8sMetadata {
-		ipc.namedPorts.Store(&npm)
-	}
-
-	// Set namedPortsChanged to false if they have not (yet) been requested.
-	// This avoids triggering policy updates if named ports changed, but no
-	// policy actually needs them. We still pre-compute the namedPorts map,
-	// so we don't have to acquire the IPCache lock in GetNamedPorts
-	if !ipc.needNamedPorts.Load() {
-		namedPortsChanged = false
-	}
```
This has been removed, where as the corresponding `Store` is still there in `GetNamedPorts` which is now dead code.

I'm a bit torn on how to tackle this (should we remove the boolean? should we keep it?). The comment here is still valid, i.e. by keeping this boolean we avoid re-triggering policies after namedPort changes if no policy ever used namedPorts. 

But as I understand now, even before this PR we were potentially triggering unnecessary policy updates if a namedPorts policy was added and then later removed (since this is never set back to false). This makes me wonder how necessary this flag here is in the first place, i.e. how expensive is it to trigger policy updates if no policy is using named ports?

bimmlerd replied:
I've pushed a fixup 56a943c93b1d7a55a212eaa4e5024fb7fb25ef14 which shows the changes needed to keep the boolean, I think it's probably worth it as it doesn't increase the code complexity much and may save some computation.

But yes, we've never set this back to false, so it might not be the most effective of optimisations.

gandro replied:
> I think it's probably worth it as it doesn't increase the code complexity much and may save some computation.

Agreed. Let's keep it to avoid further changing behavior on stable branches, especially since we backport this. I've left one comment on the fixup commit.

gandro commented on pkg/ipcache/ipcache_test.go:
```diff
@@ -604,9 +621,10 @@ func benchmarkIPCacheUpsert(b *testing.B, num int) {
 		})
+
+		b.StartTimer()
+		// We only want to measure the call to upsert.
```
Why this change?

gandro replied:
Ah, I just noticed https://github.com/cilium/cilium/pull/25007#discussion_r1172817804

Might be worth moving this change into the first commit.

gandro commented on pkg/types/portmap.go:
```diff
@@ -47,21 +49,82 @@ func (pps PortProtoSet) Equal(other PortProtoSet) bool {
-			return false
+}
+
+func (npm *namedPortMultiMap) Get(name string) (PortProtoSet, bool) {
```
This looks off. We're handing out a reference to `PortProtoSet` here, which I think allows consumers of this function to read (or even write) to the `PortProtoSet` without holding `npm.RWMutex`, which would be racy.

The good news is that I don't see any usage of this function, so we probably can just remove it?

bimmlerd replied:
Yeah - I realised the problem with it as well, and removed it form the interface but forgot to remove the implementation :facepalm: Fixing.

gandro reviewed (changes requested):
Awesome work, thanks for taking care of this! :pray: 

There is one issue around `ipc.needNamedPorts` and the externally visible behavior that I'm not sure how to address, see inline comment.

gandro commented on pkg/ipcache/ipcache.go:
```diff
@@ -419,8 +419,10 @@ func (ipc *IPCache) upsertLocked(
+		// Update the named ports reference counting, but don't cause policy recomputation of no
+		// policy uses the named ports feature.
 		namedPortsChanged = ipc.namedPorts.Update(oldK8sMeta.NamedPorts, newNamedPorts)
+		namedPortsChanged = namedPortsChanged && ipc.needNamedPorts.Load()
```
I think `deleteLocked` needs the same `ipc.needNamedPorts.Load` here as well? Otherwise we would trigger policy whenever a named port is deleted.

bimmlerd replied:
Very true. fixed and squashed.

gandro commented on pkg/types/portmap.go:
```diff
@@ -47,21 +49,75 @@ func (pps PortProtoSet) Equal(other PortProtoSet) bool {
 	}
-	return true
+	for name, port := range old {
+		if PortProtoSet(npm.m[name]).Delete(port) {
```
nit: The `PortProtoSet()` type conversion here is a no-op.

gandro commented on pkg/types/portmap.go:
```diff
@@ -47,21 +49,75 @@ func (pps PortProtoSet) Equal(other PortProtoSet) bool {
-	return true
+	for name, port := range old {
+		if PortProtoSet(npm.m[name]).Delete(port) {
+			delete(npm.m, name)
```
I don't think this is right. Even if we deleted a port, there might be other (non-conflicting) ports with the same name. We only should do a `delete(npm.m, name)` if `len(npm.m) == 0`.

Here's a unit test to demonstrate what I mean:

It fails at the end, because `http=8080:UDP` is no longer present, even though it has not been deleted.
```go

func (ds *PortsTestSuite) TestPolicyNamedPortMultiMapUpdate(c *C) {
	npm := NewNamedPortMultiMapUpdater()

	pod1PortsOld := map[string]PortProto{}
	pod1PortsNew := map[string]PortProto{
		"http": {80, uint8(u8proto.TCP)},
	}

	// Insert http=80/TCP from pod1
	changed := npm.Update(pod1PortsOld, pod1PortsNew)
	c.Assert(changed, Equals, true)

	// No changes
	changed = npm.Update(pod1PortsNew, pod1PortsNew)
	c.Assert(changed, Equals, false)

	// Assert http=80/TCP
	port, err := npm.GetNamedPort("http", uint8(u8proto.TCP))
	c.Assert(err, IsNil)
	c.Assert(port, Equals, uint16(80))

	pod2PortsOld := map[string]PortProto{}
	pod2PortsNew := map[string]PortProto{
		"http": {8080, uint8(u8proto.UDP)},
	}

	// Insert http=8080/UDP from pod2, retain http=80/TCP from pod1
	changed = npm.Update(pod2PortsOld, pod2PortsNew)
	c.Assert(changed, Equals, true)

	port, err = npm.GetNamedPort("http", uint8(u8proto.TCP))
	c.Assert(err, IsNil)
	c.Assert(port, Equals, uint16(80))
	port, err = npm.GetNamedPort("http", uint8(u8proto.UDP))
	c.Assert(err, IsNil)
	c.Assert(port, Equals, uint16(8080))

	// Delete http=80/TCP from pod1, retain http=8080/UDP from pod2
	pod1PortsOld = pod1PortsNew
	pod1PortsNew = map[string]PortProto{}

	// Delete http=80/TCP from pod1
	changed = npm.Update(pod1PortsOld, pod1PortsNew)
	c.Assert(changed, Equals, true)

	port, err = npm.GetNamedPort("http", uint8(u8proto.UDP))
	c.Assert(err, IsNil)
	c.Assert(port, Equals, uint16(8080))
}
```

bimmlerd replied:
Nice catch. Added the testcase, fixed the problem - PTAL.

gandro reviewed (changes requested):
Thanks! Starting to look good overall. I found one bug that I think does not match the expected semantics I think

joestringer reviewed (commented):
I perused this PR for a while, but I had a bit of trouble pinpointing the key algorithm changes in the code given that the main commit here contains a fair bit of refactoring as well as the logic change. It seems like you and @gandro have a handle on it, so I'm inclined to leave it be, but if it's at all possible to split refactoring and key changes into separate commits, that would help subsequent review and debugging if we ever have to go back and `git bisect` this change. I briefly tried to do this, but I could only identify that the `pkg/ipcache/ipcache_test.go` changes could be isolated into a dedicated commit, and that didn't really help with understanding the code. I'll put this down for now but shout if you think this needs more close attention.

bimmlerd committed:
ipcache: make NamedPortMultiMap an interface

In preparation of introducing reference counting for named ports, change
the NamedPortMultiMap from a concrete type to an interface. This
produces a fair bit of churn in the test code (as it assumes the type is
a map). There are no functional changes, however.

The benefit of having this type be an interface is that we can switch
it's implementing type to use reference counting with a smaller patch.
That, in turn, will allow us to avoid looking at all other pods when a
new pod is added or deleted, just for the named port bookkeeping.

Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
ipcache: switch named ports to reference counting

This commit introduces reference counting for named ports. Using the
reference counting, we know when to add or remove named ports in our
bookkeeping, without having to look through all other pods. This leads
to a significant speedup when inserting many pods, as is the case on
agent startup.

Note that this patch does _not_ restore the original optimization of not
keeping track of named ports until used in a policy. We did not find a
good way of doing so while also avoiding re-introducing the deadlock
fixed in the commit referenced below. Instead, we always do the
bookkeeping, but in a more efficient manner.

The reference counting requires the use of a RWLock, since read and
write accesses to a map must be protected. Thus, there is potential for
introducing yet another deadlock. However, we argue that this is not the
case: We do _not_ acquire any other lock while holding our mutex M. A
wait cycle/deadlock involving M must include holding M when acquiring
some other lock, however. Therefore this patch cannot introduce a
deadlock.

Fixes: 93cd67f299 (ipcache: fix potential deadlock in GetNamedPorts)
Fixes: #24987

Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
ipcache: simplify the named ports update code

With the reference counting in place, we can simplify the involved code
paths in the ipcache. There is now little point to conditionally perform
the update, as the update itself is about as expensive as checking
whether it is necessary. Therefore, opt for simpler code and always call
into NamedPortMultiMapUpdater.Update.

Indeed, this even gives a mild speedup:

goos: linux
goarch: arm64
pkg: github.com/cilium/cilium/pkg/ipcache
                      │   fix.txt   │            cleanup.txt             │
                      │   sec/op    │   sec/op     vs base               │
IPCacheUpsert10-10      48.84µ ± 5%   47.59µ ± 5%       ~ (p=0.165 n=10)
IPCacheUpsert100-10     508.3µ ± 5%   485.0µ ± 2%  -4.58% (p=0.005 n=10)
IPCacheUpsert1000-10    5.282m ± 5%   5.056m ± 2%  -4.28% (p=0.011 n=10)
IPCacheUpsert10000-10   48.63m ± 5%   44.87m ± 5%  -7.74% (p=0.009 n=10)
geomean                 1.589m        1.513m       -4.81%

Suggested-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

ldelossa commented on pkg/types/portmap.go:
```diff
@@ -49,21 +49,38 @@ func (pps PortProtoSet) Equal(other PortProtoSet) bool {
-func (npm NamedPortMultiMap) Equal(other NamedPortMultiMap) bool {
-	if len(npm) != len(other) {
+func (npm namedPortMultiMap) Equal(other NamedPortMultiMap) bool {
+	o, ok := other.(namedPortMultiMap)
```
This strikes me as a bit odd. 

You take an interface in, but you pretty much force the interface to be a concrete type for comparison. 

Is this desired? If expose a way to obtain the {name, port} tuples in a generic way, then you can avoid the forcing the a type within the interface method implementation.

joestringer replied:
Yeah, slightly odd but it looks like this is an intermediate state in the PR which is useful for transitioning to a subsequent implementation, and it's deleted by the end. There should be no risk of falsely asserting the wrong type :+1:

ldelossa replied:
Yup, reviewed this per commit, so this is even marked as outdated. Resolving

ldelossa commented on pkg/types/portmap.go:
```diff
@@ -47,38 +49,75 @@ func (pps PortProtoSet) Equal(other PortProtoSet) bool {
-func NewNamedPortMultiMap() namedPortMultiMap {
-	return make(namedPortMultiMap)
+// NamedPortMultiMapUpdater allows for mutation of the NamedPortMultiMap, and is
+// intended for ipcache internal usage.
```
`intended for ipcache internal usage` makes me wonder why this is not in `ipcache` package and why its exported?

bimmlerd replied:
Good point, fixed.

ldelossa reviewed (approved):
Overall code looks good. Just a few things that made me scratch my head while reading. Don't think either are worth blocking the PR on.

joestringer commented on pkg/ipcache/ipcache.go:
nit, commit message:

`Fixes: 93cd67f2994a618865fbcdb0294bcd9b4d44036a`

It's typically more common to use this format to give more context while still disambiguating longer hashes:

```
$ git log --oneline --pretty='Fixes: %h ("%s")' -1 93cd67f2994a618865fbcdb0294bcd9b4d44036a
Fixes: 93cd67f2994a ("ipcache: fix potential deadlock in GetNamedPorts")
```

(Personally I just put this in a special alias in my `.gitconfig`  `[pretty]` section to format this)

christarazi replied:
(Oh nice tip! I'll steal that :slightly_smiling_face:)

bimmlerd replied:
Good to know - fixed and +1 on the alias :)

joestringer reviewed (approved):
Thanks, I didn't quite understand the role of the last patch until you split it up here. LGTM :+1:

gandro commented on pkg/types/portmap.go:
```diff
@@ -47,38 +49,75 @@ func (pps PortProtoSet) Equal(other PortProtoSet) bool {
-		if otherPorts, exists := o[name]; !exists || !ports.Equal(otherPorts) {
-			return false
+	for name, port := range old {
+		namedPortsChanged = npm.m[name].Delete(port) || namedPortsChanged
```
tiny nit: I _think_ we only need to check the len if `npm.m[name].Delete(port)` returned true. I don't think there is any other condition when it would be zero.

```go
	if npm.m[name].Delete(port) {
		namedPortsChanged = true
		if len(npm.m[name]) == 0 {
			delete(npm.m, name)
		}
	}
```

(As a bonus point, it would bring back the aestethic symmetry with Add, which uses an `if`-statement instead of `||` to set `namedPortsChanged`)

bimmlerd replied:
I agree, looks better and passes tests, changed. Thanks!

gandro reviewed (approved):
Awesome! One tiny optional nit, besides the things already pointed out by the other reviewers
